openmls/src/group/mls_group/validation.rs

```diff
@@ -0,0 +1,244 @@
+//! This module contains validation functions for incoming messages
+//! as defined in https://github.com/openmls/openmls/wiki/Message-validation
+
+use std::collections::HashSet;
+
+use super::{proposals::StagedProposalQueue, *};
+
+impl MlsGroup {
+    // === Messages ===
+
+    /// Checks the following semantic validation:
+    ///  - ValSem2
+    ///  - ValSem3
+    pub fn validate_framing(&self, message: &MlsMessageIn) -> Result<(), MlsGroupError> {
+        // ValSem2
+        if message.group_id() != self.group_id() {
+            return Err(FramingValidationError::WrongGroupId.into());
+        }
+
+        // ValSem3
+        if message.epoch() != self.context().epoch() {
+            return Err(FramingValidationError::WrongEpoch.into());
+        }
+
+        Ok(())
+    }
+
+    /// Checks the following semantic validation:
+    ///  - ValSem4
+    ///  - ValSem5
+    ///  - ValSem7
+    ///  - ValSem9
+    pub fn validate_plaintext(
+        &self,
+        plaintext: &VerifiableMlsPlaintext,
+    ) -> Result<(), MlsGroupError> {
+        // ValSem4
+        let sender = plaintext.sender();
+        if sender.is_member() {
+            let sender_index = sender.to_leaf_index();
+            if sender_index > self.tree().leaf_count() || self.tree().nodes[sender_index].is_blank()
+            {
+                return Err(FramingValidationError::UnknownMember.into());
+            }
+        }
+
+        // ValSem5
```

Comment:
Here, we check more than what's advertised by ValSem5. Should the second test get its own label? It's related, but not quite the same, I think. No strong opinion, though.

Reply:
I don't understand what you mean. I added a comment to make the check more clear.

Reply:
I was just wondering if we should formally introduce the additional check that we make here: That application messages must originate from a group member.